net: dns: Fix DNS caching bug

nolson-wl opened the pull request:
The current DNS caching logic is not aware of the IP address version. If there is a cached address for a query, the caller of dns_get_addr_info() will receive that address, even if it is not the same version as the caller requested. For example:

- dns_get_addr_info() is called to resolve an IPv4 address
- the DNS client caches the IPv4 address that was received
- dns_get_addr_info() is called to resolve an IPv6 address
- the DNS client sees that there is a cached IPv4 address for the given query
- the DNS client does not check that the cached address is of the requested version (IPv6)
- the cached IPv4 address is returned from dns_get_addr_info()

This changes the DNS client to check IP address version when searching cached addresses.

jukkar commented on subsys/net/lib/dns/dns_cache.h:
```diff
@@ -94,7 +94,7 @@ int dns_cache_remove(struct dns_cache *cache, char const *query);
  */
-int dns_cache_find(struct dns_cache const *cache, const char *query, struct dns_addrinfo *addrinfo,
-		   size_t addrinfo_array_len);
+int dns_cache_find(struct dns_cache const *cache, const char *query, sa_family_t family,
```
Please add the family parameter to the doxygen comment above

jukkar commented on subsys/net/lib/dns/dns_cache.c:
```diff
@@ -97,8 +97,8 @@ int dns_cache_remove(struct dns_cache *cache, char const *query)
 	return 0;
```
Please remove the `subsys: ` from commit subject, it is redundant information and not needed. So start the commit by `net:`

jukkar commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -1356,6 +1356,9 @@ int dns_resolve_name(struct dns_resolve_context *ctx,
 	if (!ctx || !query || !cb) {
 		return -EINVAL;
 	}
+	if (type != DNS_QUERY_TYPE_A && type != DNS_QUERY_TYPE_AAAA) {
+		return -EINVAL;
+	}
```
If the query type would be `DNS_RR_TYPE_ANY`, then we could return either address. This is probably something to think in the future and not really relevant in this PR though.

jukkar commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -1414,7 +1417,8 @@ int dns_resolve_name(struct dns_resolve_context *ctx,
 try_resolve:
 #ifdef CONFIG_DNS_RESOLVER_CACHE
-	ret = dns_cache_find(&dns_cache, query, cached_info, ARRAY_SIZE(cached_info));
+	ret = dns_cache_find(&dns_cache, query, type == DNS_QUERY_TYPE_A ? AF_INET : AF_INET6,
```
Regarding my earlier ANY resource type, instead of the family (which is not really a DNS construct), we could give the resource type parameter to `dns_cache_find()` and then figure out the family there. Then it is easier to add the ANY support later.

nolson-wl commented:
@jukkar Thank you for your review and for pointing out that the DNS system may accept the ANY type in the future. I believe I have addressed the items that you noted.

jukkar reviewed (approved):
Thanks for your contribution!

rlubos commented:
From compliance:
`author email (Noah Olson <noah@wavelynxtech.com>) needs to match one of the signed-off-by entries.`
Wrong mail in the signoff?

nolson-wl committed:
net: dns: Fix DNS caching bug

The current DNS caching logic is not aware of the IP address version.
If there is a cached address for a query, the caller of
dns_get_addr_info() will receive that address, even if it is not the
same version as the caller requested. For example:

- dns_get_addr_info() is called to resolve an IPv4 address
- the DNS client caches the IPv4 address that was received
- dns_get_addr_info() is called to resolve an IPv6 address
- the DNS client sees that there is a cached IPv4 address for the
  given query
- the DNS client does not check that the cached address is of the
  requested version (IPv6)
- the cached IPv4 address is returned from dns_get_addr_info()

This changes the DNS client to check IP address version when searching
cached addresses.

Signed-off-by: Noah Olson <noah@wavelynx.com>

rlubos commented:
> > From compliance: `author email (Noah Olson <noah@wavelynxtech.com>) needs to match one of the signed-off-by entries.` Wrong mail in the signoff?
> 
> Apologies. My email address was outdated when I made the commit. I updated it before amending, but the author email is not automatically changed when amending. I have fixed the author now.

No worries, looks we're good now :+1:
